### PR TITLE
[ new ] notions of finiteness

### PR DESCRIPTION
**Pull request by Saransh-cpp**
A subset of #2017 containing minimal changes related to finiteness

### CODE REVIEW

**Comment by laMudri** [comment]
This PR on its own fails the “definitions should come with proofs” test, but maybe it's permissible with the promise of proofs in a later PR. We'll have to let the maintainers decide.

**Comment by Saransh-cpp** [comment]
Yes, that is very true. I just don't how to include the proofs without bringing in a lot more changes (which should be in separate PRs according to the discussions in #2017).

**Comment by jamesmckinna** [comment]
(Overlooked in yesterday's meeting)
Suggest at the very least that we bump this to v2.2, or even close until/unless someone comes along able to pick up the slack?

**Comment by JacquesCarette** [comment]
I put it in v2.2 as I really would like to push this through.

File: src/Relation/Nullary/Finite/Setoid.agda [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  field
+    size : ℕ
+    srj : Surjection (setoid (Fin size)) X
+
+record InjectsIntoFinitelyEnumerable (X : Setoid c ℓ) c′ ℓ′
+       : Set (c ⊔ ℓ ⊔ lsuc (c′ ⊔ ℓ′)) where
+  field
+    Apex : Setoid c′ ℓ′
+    finitelyEnumerable : FinitelyEnumerable Apex
+    inj : Injection X Apex
```
Comment (jamesmckinna) [inline src/Relation/Nullary/Finite/Setoid.agda:45]: @JacquesCarette I can *just about* take `Apex` as the name of the 'covering object' `A` in an epic situation `A → X` (L52-L54 below), but it seems less defensible here in the dual situation `X → A`. Is it 'standard' terminology? Or is there something better from eg the categorical literature?

File: src/Relation/Nullary/Finite/Setoid.agda [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  open FinitelyEnumerable finitelyEnumerable public
+
+record SurjectionFromFinitelyEnumerable (X : Setoid c ℓ) c′ ℓ′
+       : Set (c ⊔ ℓ ⊔ lsuc (c′ ⊔ ℓ′)) where
+  field
+    Apex : Setoid c′ ℓ′
+    subfinite : Subfinite Apex
+    srj : Surjection Apex X
```
Comment (jamesmckinna) [inline src/Relation/Nullary/Finite/Setoid.agda:54]: See above. I'd prefer a 'better' name, even though I appreciate it's useful to have a field name that is clearly 'standardised apart' from any potential clashes...

... ob: which, will `Apex` ever be mentioned explicitly in use of these records, or could it be declared implicit, as it's governed by the more-informative subsequent field name in each case?

**Comment by jamesmckinna** [comment]
> I put it in v2.2 as I really would like to push this through.

Comments on notation. I'd like to see this too, but as a fresh addition, we should try to get a 'convenient'/usable/memorable syntax/naming convention fixed before committing...

File: src/Relation/Nullary/Finite/Setoid.agda (hunk 2 above)
Comment (laMudri) [reply src/Relation/Nullary/Finite/Setoid.agda:54]: I suppose being able to answer the latter question is why PRs normally come with proofs. They'd give some idea as to what the relevant idioms are.

For what it's worth, I would guess that `Apex` often does have to be given and referred to explicitly. At least in basic cases, I expect that a lot of the work will go into finding the right setoid, and then backward-chaining all of the required maps.

File: src/Relation/Nullary/Finite/Setoid.agda (hunk 1 above)
Comment (JacquesCarette) [reply src/Relation/Nullary/Finite/Setoid.agda:45]: Yes, this is probably too influenced by other code / ideas, and no longer really fits here. Suggestions?

File: src/Relation/Nullary/Finite/Setoid.agda (hunk 2 above)
Comment (JacquesCarette) [reply src/Relation/Nullary/Finite/Setoid.agda:54]: I agree that PRs that introduce new ideas should not be so minimalistic. (My preference for small PRs was pushed too far here.)

File: src/Relation/Nullary/Finite/Setoid.agda (hunk 1 above)
Comment (jamesmckinna) [reply src/Relation/Nullary/Finite/Setoid.agda:45]: I hunted around, and unfortunately the literature uses 'generic' names such as `X`, `Y`/`A`, `B` when discussing things as morphisms. For embedding/extension, the domain of an injection is sometimes called `I`, and of a projection as `P`, but these don't seem quite satisfactory, either?

In each case, though, the relevant object is (suitably) enumerable, so could it be called `Enum`, or even `Enumeration`?

Comment (JacquesCarette) [reply src/Relation/Nullary/Finite/Setoid.agda:45]: `Enumeration` > `Enum` > `E`.

**Comment by MatthewDaggitt** [comment]
I've converted this to a draft. @JacquesCarette when it's ready to review properly, turn it back again :+1:

**Comment by JacquesCarette** [comment]
We've decided to close this for now, in favour of the "project issue #2511 " as it's a bigger deal.

**Pull request closed without merging**